File: 4_Behavioral_Design_Patterns/4.03_Interpreter/Python_Snippet/Interpreter.py
class Context:
    def __init__(self):
        self.variables = {}

    def get_value(self, var_name):
        return self.variables.get(var_name, 0)

    def set_value(self, var_name, value):
        self.variables[var_name] = value


class Expression:
    def interpret(self, context):
        pass


class Number(Expression):
    def __init__(self, value):
        self.value = value

    def interpret(self, context):
        return self.value


class Plus(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def interpret(self, context):
        return self.left.interpret(context) + self.right.interpret(context)


class Minus(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def interpret(self, context):
        return self.left.interpret(context) - self.right.interpret(context)


class Multiply(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def interpret(self, context):
        return self.left.interpret(context) * self.right.interpret(context)


if __name__ == "__main__":
    context = Context()
    context.set_value("x", 3)
    context.set_value("y", 5)

    # Interpret the expression: x + y * 2
    expression = Plus(Number(context.get_value("x")), Multiply(Number(context.get_value("y")), Number(2)))
    result = expression.interpret(context)
    print("Result:", result)  # Output: 13

File: 4_Behavioral_Design_Patterns/4.03_Interpreter/Java_Snippet/Interpreter.java
import java.util.HashMap;
import java.util.Map;

// Context class to hold variables and their values
class Context {
    private Map<String, Integer> variables;

    public Context() {
        this.variables = new HashMap<>();
    }

    public int getValue(String varName) {
        return variables.getOrDefault(varName, 0);
    }

    public void setValue(String varName, int value) {
        variables.put(varName, value);
    }
}

// Abstract Expression class
abstract class Expression {
    public abstract int interpret(Context context);
}

// Terminal Expression representing a number
class Number extends Expression {
    private int value;

    public Number(int value) {
        this.value = value;
    }

    @Override
    public int interpret(Context context) {
        return value;
    }
}

// Non-terminal Expression representing addition
class Plus extends Expression {
    private Expression left;
    private Expression right;

    public Plus(Expression left, Expression right) {
        this.left = left;
        this.right = right;
    }

    @Override
    public int interpret(Context context) {
        return left.interpret(context) + right.interpret(context);
    }
}

// Non-terminal Expression representing subtraction
class Minus extends Expression {
    private Expression left;
    private Expression right;

    public Minus(Expression left, Expression right) {
        this.left = left;
        this.right = right;
    }

    @Override
    public int interpret(Context context) {
        return left.interpret(context) - right.interpret(context);
    }
}

// Non-terminal Expression representing multiplication
class Multiply extends Expression {
    private Expression left;
    private Expression right;

    public Multiply(Expression left, Expression right) {
        this.left = left;
        this.right = right;
    }

    @Override
    public int interpret(Context context) {
        return left.interpret(context) * right.interpret(context);
    }
}

public class Interpreter {
    public static void main(String[] args) {
        Context context = new Context();
        context.setValue("x", 3);
        context.setValue("y", 5);

        // Interpret the expression: x + y * 2
        Expression expression = new Plus(
                new Number(context.getValue("x")),
                new Multiply(
                        new Number(context.getValue("y")),
                        new Number(2)));
        int result = expression.interpret(context);
        System.out.println("Result: " + result); // Output: 13
    }
}
